keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1176 @@
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### Out-of-tree Implementation
+
+We could implement NodeMainentance API out-of-tree first as a CRD with node maintenance controller.
+
+One of the problems is that it would be difficult to get real word adoption and thus important
+feedback on this feature. This is mainly due to the requirement that the feature be implemented
+and integrated with multiple components to observe the benefits. And those components are both
+admin and application developer facing.
+
+There is a [Node Maintenance Operator](https://github.com/medik8s/node-maintenance-operator) project
+that provides a similar API and has some adoption. But it is not at a level where applications could
+depend on this API to be present in the cluster. So it doesn't make that much sense to implement
+the app migration logic as it cannot be applied everywhere. As shown in the motivation part
+of this KEP, there is a big appetite for having a unified and stable API that could be used by
+everyone to implement the new capabilities.
+
+### Use a Node Object Instead of Introducing a New NodeMaintenance API
+As an alternative, it would be possible to signal the node maintenance by marking the node object
+instead of introducing a new API. But it is probably better to decouple this from the node for
+extensibility reasons. As we can see, the `kubectl drain` logic is a bit complex, and it may be
+possible to move this logic to a controller in the future and make the node maintenance purely
+declarative. It also helps to decouple RBAC permissions from the node object.
+
```

Comment:
Another benefit IMO: you could have two or more different grounds for node maintenance, for a given node, that overlap in time.

Reply:
+1, I have added this point there